public/app/features/alerting/unified/components/rules/StateHistory.tsx

```diff
@@ -44,31 +54,106 @@ const StateHistory: FC<RuleStateHistoryProps> = ({ alertId }) => {
     { id: 'timestamp', label: 'Time', size: 'max-content', renderCell: renderTimestampCell },
   ];
 
-  const items: StateHistoryRow[] = result
-    .reduce((acc: StateHistoryRowItem[], item, index) => {
-      acc.push({
-        id: String(item.id),
-        state: item.newState,
-        text: item.text,
-        data: item.data,
-        timestamp: item.updated,
-      });
-
-      // if the preceding state is not the same, create a separate state entry – this likely means the state was reset
-      if (!hasMatchingPrecedingState(index, result)) {
-        acc.push({ id: uniqueId(), state: item.prevState });
-      }
-
-      return acc;
-    }, [])
-    .map((historyItem) => ({
-      id: historyItem.id,
-      data: historyItem,
-    }));
-
-  return <DynamicTable cols={columns} items={items} />;
+  const items: StateHistoryRowItem[] = result.reduce((acc: StateHistoryRowItem[], item) => {
```

Comment:
What do you think about moving this `reduce` to the `useManagedAlertStateHistory` hook?

Reply:
I agree, in case we are only going to use this hook for this particular use case.

Reply:
I usually try to keep the data structures in hooks as close to the API response as possible but seeing how we're not using this hook anywhere else it's sensible to move this to the hook – additionally there's no more reason for this to be a `reduce` function; we previously appended some items but now it's just a plain `map`.

Reply:
I'll refactor this in to a separate function (and write some tests for it), that way we can have the hook that matches the API response and a function to transform the `result` from the `AsyncRequestState` where we need to

Reply:
💯 I was about to say that, (about the map) but I thought you were used to use reduce instead.

Reply:
this refactor makes a lot of sense @gillesdemey 🙌